File: api/src/main/java/org/apache/gravitino/authorization/RoleChange.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.authorization;

import java.util.Objects;
import org.apache.gravitino.annotation.Evolving;

/** The RoleChange interface defines the public API for managing roles in an authorization. */
@Evolving
public interface RoleChange {
  /**
   * Create a RoleChange to add a securable object into a role.
   *
   * @param roleName The role name.
   * @param securableObject The securable object.
   * @return return a RoleChange for the added securable object.
   */
  static RoleChange addSecurableObject(String roleName, SecurableObject securableObject) {
    return new AddSecurableObject(roleName, securableObject);
  }

  /**
   * Create a RoleChange to remove a securable object from a role.
   *
   * @param roleName The role name.
   * @param securableObject The securable object.
   * @return return a RoleChange for the added securable object.
   */
  static RoleChange removeSecurableObject(String roleName, SecurableObject securableObject) {
    return new RemoveSecurableObject(roleName, securableObject);
  }

  /**
   * Update a securable object RoleChange.
   *
   * @param roleName The role name.
   * @param securableObject The securable object.
   * @param newSecurableObject The new securable object.
   * @return return a RoleChange for the update securable object.
   */
  static RoleChange updateSecurableObject(
      String roleName, SecurableObject securableObject, SecurableObject newSecurableObject) {
    return new UpdateSecurableObject(roleName, securableObject, newSecurableObject);
  }

  /** A AddSecurableObject to add a securable object to a role. */
  final class AddSecurableObject implements RoleChange {
    private final String roleName;
    private final SecurableObject securableObject;

    private AddSecurableObject(String roleName, SecurableObject securableObject) {
      this.roleName = roleName;
      this.securableObject = securableObject;
    }

    /**
     * Returns the role name to be added.
     *
     * @return return a role name.
     */
    public String getRoleName() {
      return this.roleName;
    }

    /**
     * Returns the securable object to be added.
     *
     * @return return a securable object.
     */
    public SecurableObject getSecurableObject() {
      return this.securableObject;
    }

    /**
     * Compares this AddSecurableObject instance with another object for equality. The comparison is
     * based on the added securable object to a role.
     *
     * @param o The object to compare with this instance.
     * @return true if the given object represents the same as the add securable object; false
     *     otherwise.
     */
    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      AddSecurableObject that = (AddSecurableObject) o;
      return roleName.equals(that.roleName) && securableObject.equals(that.securableObject);
    }

    /**
     * Generates a hash code for this AddSecurableObject instance. The hash code is based on the add
     * securable object.
     *
     * @return A hash code value for this add securable object operation.
     */
    @Override
    public int hashCode() {
      return Objects.hash(roleName, securableObject);
    }

    /**
     * Returns a string representation of the AddSecurableObject instance. This string format
     * includes the class name followed by the add securable object operation.
     *
     * @return A string representation of the AddSecurableObject instance.
     */
    @Override
    public String toString() {
      return "ADDSECURABLEOBJECT " + roleName + " " + securableObject;
    }
  }

  /** A RemoveSecurableObject to remove a securable object from a role. */
  final class RemoveSecurableObject implements RoleChange {
    private final String roleName;
    private final SecurableObject securableObject;

    private RemoveSecurableObject(String roleName, SecurableObject securableObject) {
      this.roleName = roleName;
      this.securableObject = securableObject;
    }

    /**
     * Returns the role name.
     *
     * @return return a role name.
     */
    public String getRoleName() {
      return roleName;
    }

    /**
     * Returns the securable object to be added.
     *
     * @return return a securable object.
     */
    public SecurableObject getSecurableObject() {
      return this.securableObject;
    }

    /**
     * Compares this RemoveSecurableObject instance with another object for equality. The comparison
     * is based on the added securable object to a role.
     *
     * @param o The object to compare with this instance.
     * @return true if the given object represents the same as remove securable object; false
     *     otherwise.
     */
    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      RemoveSecurableObject that = (RemoveSecurableObject) o;
      return roleName.equals(that.roleName) && securableObject.equals(that.securableObject);
    }

    /**
     * Generates a hash code for this RemoveSecurableObject instance. The hash code is based on the
     * added securable object.
     *
     * @return A hash code value for this adds securable object operation.
     */
    @Override
    public int hashCode() {
      return Objects.hash(roleName, securableObject);
    }

    /**
     * Returns a string representation of the RemoveSecurableObject instance. This string format
     * includes the class name followed by the add securable object operation.
     *
     * @return A string representation of the RemoveSecurableObject instance.
     */
    @Override
    public String toString() {
      return "REMOVESECURABLEOBJECT " + roleName + " " + securableObject;
    }
  }

  /**
   * A UpdateSecurableObject is to update securable object's privilege from a role. <br>
   * The securable object's metadata entity must be the same as new securable object's metadata
   * entity. <br>
   * The securable object's privilege must be different from new securable object's privilege. <br>
   */
  final class UpdateSecurableObject implements RoleChange {
    private final String roleName;
    private final SecurableObject securableObject;
    private final SecurableObject newSecurableObject;

    private UpdateSecurableObject(
        String roleName, SecurableObject securableObject, SecurableObject newSecurableObject) {
      if (!securableObject.fullName().equals(newSecurableObject.fullName())) {
        throw new IllegalArgumentException(
            "The securable object's metadata entity must be same as new securable object's metadata entity.");
      }
      if (securableObject.privileges().equals(newSecurableObject.privileges())) {
        throw new IllegalArgumentException(
            "The securable object's privilege must be different as new securable object's privilege.");
      }
      this.roleName = roleName;
      this.securableObject = securableObject;
      this.newSecurableObject = newSecurableObject;
    }

    /**
     * Returns the role name.
     *
     * @return return a role name.
     */
    public String getRoleName() {
      return roleName;
    }

    /**
     * Returns the securable object to be updated.
     *
     * @return return a securable object.
     */
    public SecurableObject getSecurableObject() {
      return this.securableObject;
    }

    /**
     * Returns the new securable object.
     *
     * @return return a securable object.
     */
    public SecurableObject getNewSecurableObject() {
      return this.newSecurableObject;
    }

    /**
     * Compares this UpdateSecurableObject instance with another object for equality. The comparison
     * is based on the old securable object and new securable object.
     *
     * @param o The object to compare with this instance.
     * @return true if the given object represents the same update securable object; false
     *     otherwise.
     */
    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      UpdateSecurableObject that = (UpdateSecurableObject) o;
      return roleName.equals(that.roleName)
          && securableObject.equals(that.securableObject)
          && newSecurableObject.equals(that.newSecurableObject);
    }

    /**
     * Generates a hash code for this UpdateSecurableObject instance. The hash code is based on the
     * old securable object and new securable object.
     *
     * @return A hash code value for this update securable object operation.
     */
    @Override
    public int hashCode() {
      return Objects.hash(roleName, securableObject, newSecurableObject);
    }

    /**
     * Returns a string representation of the UpdateSecurableObject instance. This string format
     * includes the class name followed by the add securable object operation.
     *
     * @return A string representation of the RemoveSecurableObject instance.
     */
    @Override
    public String toString() {
      return "UPDATESECURABLEOBJECT " + roleName + " " + securableObject + " " + newSecurableObject;
    }
  }
}

File: clients/client-python/gravitino/api/authorization/role_change.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from abc import ABC
from dataclasses import dataclass, field

from dataclasses_json import config

from gravitino.api.authorization.securable_objects import SecurableObject
from gravitino.utils.precondition import Precondition


class RoleChange(ABC):
    """The RoleChange interface defines the public API for managing roles in an authorization."""

    @staticmethod
    def add_securable_object(
        role_name: str,
        securable_object: SecurableObject,
    ) -> RoleChange.AddSecurableObject:
        """
        Create a RoleChange to add a securable object into a role.

        Args:
            role_name (str): The role name.
            securable_object (SecurableObject): The securable object.

        Returns:
            RoleChange.AddSecurableObject: return a RoleChange for the added securable object.
        """
        return RoleChange.AddSecurableObject(role_name, securable_object)

    @staticmethod
    def remove_securable_object(
        role_name: str,
        securable_object: SecurableObject,
    ) -> RoleChange.RemoveSecurableObject:
        """
        Create a RoleChange to remove a securable object from a role.

        Args:
            role_name (str): The role name.
            securable_object (SecurableObject): The securable object.

        Returns:
            RoleChange.RemoveSecurableObject: return a RoleChange for the added securable object.
        """
        return RoleChange.RemoveSecurableObject(role_name, securable_object)

    @staticmethod
    def update_securable_object(
        role_name: str,
        securable_object: SecurableObject,
        new_securable_object: SecurableObject,
    ) -> RoleChange.UpdateSecurableObject:
        """
        Update a securable object RoleChange.

        Args:
            role_name (str): The role name.
            securable_object (SecurableObject): The securable object.
            new_securable_object (SecurableObject): The new securable object.

        Returns:
            RoleChange.UpdateSecurableObject: return a RoleChange for the update-securable object.
        """
        return RoleChange.UpdateSecurableObject(
            role_name, securable_object, new_securable_object
        )

    @dataclass(frozen=True, eq=True)
    class AddSecurableObject:
        """A AddSecurableObject to add a securable object to a role."""

        _role_name: str = field(metadata=config(field_name="role_name"))
        _securable_object: SecurableObject = field(
            metadata=config(field_name="securable_object")
        )

        @property
        def role_name(self) -> str:
            """
            Returns the role name to be added.

            Returns:
                str: The role name to be added.
            """
            return self._role_name

        @property
        def securable_object(self) -> SecurableObject:
            """
            Returns the securable object to be added.

            Returns:
                SecurableObject: The securable object to be added.
            """
            return self._securable_object

        def __str__(self) -> str:
            """
            Returns a string representation of the AddSecurableObject instance. This string format
            includes the class name followed by the add securable object operation.

            Returns:
                str: A string representation of the AddSecurableObject instance.
            """
            return f"ADDSECURABLEOBJECT {self._role_name} + {self._securable_object}"

    @dataclass(frozen=True, eq=True)
    class RemoveSecurableObject:
        """A RemoveSecurableObject to remove a securable object from a role."""

        _role_name: str = field(metadata=config(field_name="role_name"))
        _securable_object: SecurableObject = field(
            metadata=config(field_name="securable_object")
        )

        @property
        def role_name(self) -> str:
            """
            Returns the role name to be removed.

            Returns:
                str: The role name to be removed.
            """
            return self._role_name

        @property
        def securable_object(self) -> SecurableObject:
            """
            Returns the securable object to be removed.

            Returns:
                SecurableObject: The securable object to be removed.
            """
            return self._securable_object

        def __str__(self) -> str:
            """
            Returns a string representation of the RemoveSecurableObject instance. This string format
            includes the class name followed by the add securable object operation.

            Returns:
                str: A string representation of the RemoveSecurableObject instance.
            """
            return f"REMOVESECURABLEOBJECT {self._role_name} + {self._securable_object}"

    @dataclass(frozen=True, eq=True)
    class UpdateSecurableObject:
        """A UpdateSecurableObject is to update securable object's privilege from a role.
        The securable object's metadata entity must be the same as new securable object's metadata
        entity.
        The securable object's privilege must be different from new securable object's privilege.
        """

        _role_name: str = field(metadata=config(field_name="role_name"))
        _securable_object: SecurableObject = field(
            metadata=config(field_name="securable_object")
        )
        _new_securable_object: SecurableObject = field(
            metadata=config(field_name="new_securable_object")
        )

        def __post_init__(self) -> None:
            Precondition.check_argument(
                self._securable_object.full_name()
                == self._new_securable_object.full_name(),
                "The securable object's metadata entity must be same as new securable object's metadata entity.",
            )

            Precondition.check_argument(
                self._securable_object.privileges()
                != self._new_securable_object.privileges(),
                "The securable object's privilege must be different as new securable object's privilege.",
            )

        @property
        def role_name(self) -> str:
            """
            Returns the role name to be updated.

            Returns:
                str: The role name to be updated.
            """
            return self._role_name

        @property
        def securable_object(self) -> SecurableObject:
            """
            Returns the securable object to be updated.

            Returns:
                SecurableObject: The securable object to be updated.
            """
            return self._securable_object

        @property
        def new_securable_object(self) -> SecurableObject:
            """
            Returns the new securable object.

            Returns:
                SecurableObject: return a securable object.
            """
            return self._new_securable_object

        def __str__(self) -> str:
            """
            Returns a string representation of the UpdateSecurableObject instance. This string format
            includes the class name followed by the add securable object operation._summary_

            Returns:
                str: A string representation of the RemoveSecurableObject instance.
            """
            return (
                f"UPDATESECURABLEOBJECT "
                f"{self._role_name} {self._securable_object} {self._new_securable_object}"
            )
